docs/spec_draft.md

```diff
@@ -300,6 +316,148 @@ the IEEE-754 specification. For boolean element type, the behavior is same as
+   * Number of groups formed: G.
+   * Size of each group: S * P.
+   * Interpretation of `replica_groups`: contains replica ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     if len(replica_groups) == 0:
+       replica_groups = all_partions_ids
+
+     for g in replica_groups:
+       for rid in g:
+         for pid in all_partions_ids:
+           group += ei(rid, pid)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[0, 2], [1, 3]]` and the
+      `groups` formed: `{ei(0, 0), ei(0, 1), ei(2, 0), ei(2, 1)} and {ei(1, 0), ei(1, 1), ei(3, 0), ei(3, 1)}`.
+
+3. If `channel_handle` is provided and `use_global_device_ids` $=$ `true`
+   * Number of groups formed: G.
+   * Size of each group: S.
+   * Interpretation of `replica_groups`: contains global device ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     for g in replica_groups:
+       for global_device_id in g:
+           group += ei(global_device_id / P, global_device_id % P)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[4, 5], [6, 7]]` and the
+      `groups` formed: `{ei(2, 0), ei(2, 1)} and {ei(3, 0), ei(3, 1)}`.
+
+For each `group` $\in$ `groups`, the `result`, to be visible by each member of
+`group`, is given by `concatenate(operands, all_gather_dim)`, where
+`operands` = { `operand` w.r.t each `ei(rid, pid)` following the same order in
+`group` }.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
+| `all_gather_dim`        | constant of type `si64`                                                                                                                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64`                                                                                                  |
+| `channel_handle`        | constant of type struct { `channel-id`: `si64`, `channel-type`: enum of `CHANNEL_TYPE_INVALID`, `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, `HOST_TO_DEVICE` } |
+| `use_global_device_ids` | constant of type `boolean`                                                                                                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
```

Comment:
Could you reorder the constraints to follow the order in Inputs

Reply:
The constraints are ordered as per Inputs and Outputs. The confusing part might be C5-7, but they are related to result. Let me know?

Reply:
Yes I was referring to (C5) and (C6). I think by convention, (C5) and (C6) were the first ones listed on the constraints. Lmk if you have other reasoning:

Proposed change:
C5 -> C1
C6 -> C2
(don't have strong preference on the below):
C2 -> C3 This is a constraint on the operand using all_gather_dim
C1 -> C4
C3 -> C5
C4 -> C6
C7 -> C7